### PR TITLE
Add Knowledge Gradient candidates functions

### PR DESCRIPTION
**Pull request by alxhslm**
## Motivation

Knowledge gradient acquisition functions can perform better on certain problems since they have lookahead unlike myoptic acquisition functions such as Expected Improvement.

## Description of the changes
I've added 2 new candidates functions:
1. ` qkg_candidates_func` for single objective problems
2. `qhvkg_candidates_func` for multi-objective problems

Although the user could implement these themselves, it is non-trivial so it makes sense to add them IMO.

The tests have also been updated to test these acquisition functions.

### CODE REVIEW

File: optuna_integration/botorch/botorch.py [hunk 1]
```diff
@@ -68,6 +74,11 @@ def _get_sobol_qmc_normal_sampler(num_samples: int) -> SobolQMCNormalSampler:
     from botorch.acquisition.analytic import LogConstrainedExpectedImprovement
     from botorch.acquisition.analytic import LogExpectedImprovement
 
+with try_import() as _imports_qhvkg:
+    from botorch.acquisition.multi_objective.hypervolume_knowledge_gradient import (
```
Comment (alxhslm) [inline optuna_integration/botorch/botorch.py:78]: This was only added in a relatively recent version of BoTorch in pytorch/botorch#1950.

File: optuna_integration/botorch/botorch.py [hunk 2]
```diff
@@ -662,6 +673,151 @@ def qparego_candidates_func(
...
+        options={"batch_limit": 5, "maxiter": 200},
+        sequential=True,
+    )
+
+    candidates = unnormalize(candidates.detach(), bounds=bounds)
+
+    return candidates
+
+
+@experimental_func("3.7.0")
```
Comment (alxhslm) [inline optuna_integration/botorch/botorch.py:744]: I set this to the next minor version, which I assume is the correct thing to do.

File: optuna_integration/botorch/botorch.py [hunk 3]
```diff
@@ -662,6 +674,158 @@ def qparego_candidates_func(
...
+
+    ref_point = train_obj.min(dim=0).values - 1e-8
+
+    acqf = qHypervolumeKnowledgeGradient(
+        model=model,
+        ref_point=ref_point,
+        num_fantasies=16,
+        X_pending=pending_x,
+        objective=objective,
+        sampler=ListSampler(
```
Comment (alxhslm) [inline optuna_integration/botorch/botorch.py:803]: We set the sampler explicitly to support `botorch`==0.9.5  (replicates the changes from pytorch/botorch#2160).

**Comment by nabenabe0928** [comment]
@alxhslm Thank you for this PR!
@eukaryo Could you review this PR?

File: optuna_integration/botorch/botorch.py [hunk 4]
```diff
@@ -662,6 +674,158 @@ def qparego_candidates_func(
...
+        options={"batch_limit": 8, "maxiter": 200},
+        sequential=True,
+    )
+
+    candidates = unnormalize(candidates.detach(), bounds=bounds)
+
+    return candidates
+
+
+@experimental_func("3.7.0")
```
Comment (eukaryo) [inline optuna_integration/botorch/botorch.py:745]:
```suggestion
@experimental_func("4.0.0")
```

**Review by eukaryo — COMMENTED** [review]
Sorry for my late reply. Almost LGTM. PTAL.

File: optuna_integration/botorch/botorch.py (hunk 4 above)
Comment (alxhslm) [reply optuna_integration/botorch/botorch.py]: Thanks for spotting this! I've updated accordingly.

**Review by eukaryo — APPROVED** [review]
LGTM!
